### PR TITLE
j9gc_createJavaLangString() check for ASCII properly in Unicode

### PR DESCRIPTION
**Pull request by pshipton**
Fixes https://github.com/eclipse/openj9/issues/11650

Signed-off-by: Peter Shipton <Peter_Shipton@ca.ibm.com>

### CODE REVIEW

**Comment by pshipton** [comment]
jenkins test sanity win jdk8,jdk11

**Comment by pshipton** [comment]
The Windows Java 8 build hung for no good reason.
https://ci.eclipse.org/openj9/job/Build_JDK8_x86-64_windows_Personal/1040/

jenkins test sanity win jdk8

**Comment by pshipton** [comment]
I've confirmed this change fixes the user test case.

**Comment by pshipton** [comment]
Just realized I missed a spot to use `isUnicode`, so pushing an update.
The original JDK11 PR testing passed https://ci.eclipse.org/openj9/job/Test_openjdk11_j9_sanity.functional_x86-64_windows_Personal/605/
The running JDK8 PR testing is https://ci.eclipse.org/openj9/job/PullRequest-OpenJ9/5193/

File: runtime/gc_base/StringTable.cpp [hunk 1]
```diff
@@ -533,9 +533,16 @@ j9gc_createJavaLangString(J9VMThread *vmThread, U_8 *data, UDATA length, UDATA s
...
+	bool isUnicode = J9_ARE_ANY_BITS_SET(stringFlags, J9_STR_UNICODE);
 
-	if (VM_VMHelpers::isUTF8ASCII(data, length)) {
-		stringFlags |= J9_STR_ASCII;
+	if (isUnicode && IS_STRING_COMPRESSION_ENABLED_VM(vm)) {
+		if (VM_VMHelpers::isUnicodeASCII((const U_16 *)data, length / 2)) {
+			stringFlags |= J9_STR_ASCII;
+		}
+	} else {
+		if (VM_VMHelpers::isUTF8ASCII(data, length)) {
```
Comment (gacholio) [inline runtime/gc_base/StringTable.cpp:543]: This is still possibly called when isUnicode is true, due to the compound condition in the if above.

Comment (pshipton) [reply runtime/gc_base/StringTable.cpp:543]: Oops. Fixed.

**Review by gacholio — APPROVED** [review]

**Comment by gacholio** [comment]
jenkins test sanity win jdk8,jdk11

File: runtime/gc_base/StringTable.cpp [hunk 2]
```diff
@@ -533,9 +533,16 @@ j9gc_createJavaLangString(J9VMThread *vmThread, U_8 *data, UDATA length, UDATA s
 	j9object_t result = NULL;
 	j9object_t charArray = NULL;
 	UDATA allocateFlags = J9_GC_ALLOCATE_OBJECT_NON_INSTRUMENTABLE;
+	bool isUnicode = J9_ARE_ANY_BITS_SET(stringFlags, J9_STR_UNICODE);
 
-	if (VM_VMHelpers::isUTF8ASCII(data, length)) {
-		stringFlags |= J9_STR_ASCII;
+	if (isUnicode) {
+		if (IS_STRING_COMPRESSION_ENABLED_VM(vm) && VM_VMHelpers::isUnicodeASCII((const U_16 *)data, length / 2)) {
```
Comment (gacholio) [inline runtime/gc_base/StringTable.cpp:539]: Is it correct to only tag ASCII when compression is enabled? Maybe just an optimization?

This code could use some reworking in a separate PR.

Comment (pshipton) [reply runtime/gc_base/StringTable.cpp:539]: It's an optimization. I only found the ASCII tag used when compression is enabled.

**Comment by pshipton** [comment]
Note the PR testing passed, but the build timed out before it found an available Windows machine to update the status.

File: runtime/gc_base/StringTable.cpp (hunk 2 above)
Comment (gacholio) [reply runtime/gc_base/StringTable.cpp:539]: I'll look into cleaning this next week.

**Pull request merged**